backend/app/adapter/google/identityprovider.go

```diff
@@ -0,0 +1,75 @@
+package google
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"short/app/usecase/service"
+
+	"github.com/byliuyang/app/fw"
+)
+
+const (
+	authorizationAPI     = "https://accounts.google.com/o/oauth2/v2/auth?"
+	accessTokenAPI       = "https://www.googleapis.com/oauth2/v3/token"
+	grantType            = "authorization_code"
+	scope                = "https://www.googleapis.com/auth/userinfo.email"
+	accessType           = "offline"
+	includeGrantedScopes = "true"
+	responseType         = "code"
+)
+
+type accessTokenResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+}
+
+var _ service.IdentityProvider = (*IdentityProvider)(nil)
+
+// IdentityProvider represents Google OAuth service.
+type IdentityProvider struct {
+	clientID     string
+	clientSecret string
+	http         fw.HTTPRequest
+	redirectURI  string
+}
+
+// GetAuthorizationURL retrieves the URL of Google sign in page.
+func (g IdentityProvider) GetAuthorizationURL() string {
+	clientID := g.clientID
+	redirectURI := g.redirectURI
+	return fmt.Sprintf("%s&client_id=%s&redirect_uri=%s&scope=%s&access_type=%s&include_granted_scopes=%s&response_type=%s",
```

Comment:
Let's use "url" package to build the sign in link: https://golang.org/src/net/url/example_test.go

Reply:
changed accordingly